### ISSUE TITLE
The role of manifest.json requires clarification.

### ISSUE THREAD

**Issue by serefarikan** [devcontainers/images#877]
Apologies for the rather demanding reading title; I wanted to use a title more suitable for an issue.

I have now read [README](https://github.com/devcontainers/images/blob/main/build/README.md) a number of times, trying to piece things together. I am using the image sources under the `src` directory to create devcontainer images with various customisations. These image source(s in folders)  seem to work just fine with vs code's devcontainers support. 

Except I am failing to understand whether or not `manifest.json` has any effect on the contents of the actual image that is created. My understanding is that this file is only used for publishing to MCR, but there are statements in the README that confuse me such as:

> The manifest.json file dictates how the build process should behave as [described above](https://github.com/devcontainers/images/blob/main/build/README.md#setting-up-a-container-to-be-built). In this case, devcontainer.json points to Dockerfile used to generate the actual image.

I fail to understand what "how the build process should behave" means here. It is even more confusing trying to understand what `manifest.json` has to do with `devcontainer.json` pointing at the Dockerfile (and where in the build process that pointing happens?). 

If someone could briefly clarify whether or not manifest.json actually changes the contents (packages, config etc) of the built docker image I would be grateful. Is it safe for me to delete this file and hijack .devcontainers folder only to build my own images?  

In other words, I cannot build the big picture from the README, some help would be much appreciated :)

**Comment by eljog** [comment]
Thank you for sharing your question.

To provide an example of how the defintions inside manifest.json affects the image build process, let us take a look at the `python` image.

Tags for this image uses the VARIANT property:
https://github.com/devcontainers/images/blob/722d41d4a36ba68ea300beb2c509f6a74a6e8c2d/src/python/manifest.json#L76-L78

Following are the variants defined for this image:
https://github.com/devcontainers/images/blob/722d41d4a36ba68ea300beb2c509f6a74a6e8c2d/src/python/manifest.json#L3-L18

Based on these definitions the build process builds and deploy multiple images (one for each variant), as you can see in MCR https://mcr.microsoft.com/en-us/product/devcontainers/python/tags

You can read more about this [here in the README](https://github.com/devcontainers/images/blob/main/build/README.md#the-variants-property)

**Comment by serefarikan** [comment]
thanks for taking the time to respond @eljog . What I'm trying to confirm is that manifest.json is meant to be consumed by a specific image building pipeline, which perfoms actions such as building and deploying multiple images, but this is a mechanism **on top of the devcontainer spec and its tooling** . I.e. if I delete the manifest.json and use .devcontainer folder, I could still get a devcontainer image built, say by vs code. I would not have broken packages in the produced image or missing dependencies etc. 

In other words, my understanding is that manifest.json is not part of the devcontainers spec, but it is a meta artefact used in a specific build pipeline (from Microsoft). Is that correct?

**Comment by eljog** [comment]
That is correct.
You can build an image locally from sources using the config inside the `.devcontainer` directory.
Typically you don't need to build the image from source, instead you can use the images that are built and published in MCR https://mcr.microsoft.com/en-us/catalog?search=devcontainers

Devcontainer templates can be a good starting point to build your development environment using these images: https://containers.dev/templates

**Comment by serefarikan** [comment]
Brilliant. Thanks again. I saw the templates page but completely forgot about it. Closing the issue now that it is clear what manifest.json does.
